README
tweet is a very simple command-line tweeter. That is, it allows you to
send a twitter update to a given user account. It doesn't do anything
else. Well, actually, it also allows you to follow or unfollow accounts as
well as block and unblock users.  So there's that.  Also, this tool
suffers from feeping creatures and will (or already has) probably
spawn(ed) additional features making it not all that very simple anymore.

Requirements:
	- python (presumably > 2.5)
	- py-tweepy (http://code.google.com/p/tweepy/)

Installation:
	- install -c -m 755 src/tweet.py /somewhere/in/your/path/tweet
	- install -c -m 444 doc/tweet.1 /somewhere/in/your/manpath/man1/tweet.1

Questions/comments:
	jschauma@netmeister.org
	https://www.netmeister.org/apps/twitter/tweet/

---

```
NAME
     tweet -- send a twitter update

SYNOPSIS
     tweet [-hit] [-B user] [-F user] [-a id] [-b user] [-c file] [-d id]
	   [-f user] [-l id] [-m media] [-r id] -u user

DESCRIPTION
     tweet is a very simple tool to send a twitter update to the given account
     - nothing else.  The message to send is read from STDIN; the user account
     to update needs to be configured in the file ~/.tweetrc.

OPTIONS
     The following options are supported by tweet:

     -B user   Unblock the given user.	Can be given multiple times to unblock
	       multiple users.

     -F user   Unfollow the given user.	 Can be given multiple times to unfol-
	       low multiple users.

     -a id     Answer (ie reply) to the given id.  That is, let the API sort
	       the new message as a reply to this message.

     -b user   Block the given user.  Can be given multiple times to block
	       multiple users.

     -c file   Read the configuration from the given file.

     -d id     Delete the given message.  Can be given multiple times to
	       delete multiple messages.

     -h	       Print a short usage statement and exit.

     -i	       Print the tweet ID of any new tweets.

     -f user   Follow the given user.  Can be given multiple times to follow
	       multiple users.

     -l id     Like the given tweet.  Can be given multiple times to like mul-
	       tiple tweets.

     -m media  Upload the given media (picture or video).

     -r id     Retweet the tweet with the given id.  Can be given multiple
	       times to retweet multiple messages.

     -t	       Truncate messages that are too long.

     -u user   Update the given account.

DETAILS
     tweet requires you to specify the username of the twitter account to
     which to post the update.	After that, it will read input from STDIN.

     If the configuration file ~/.tweetrc contains information for exactly one
     username, then tweet will update that user's status.  If the file does
     not contain information for the user given on the command-line, then
     tweet will attempt to retrieve the access information for that account
     and ask the user to update the configuration file.

     If the resulting message contains 280 or fewer characters, tweet will
     post the message directly, otherwise, it will complain and exit (unless
     the -t flag is given).

     If the -r flag is given, tweet will not read anything from STDIN and sim-
     ply retweet the messages with the given ID(s).

TWITTER AUTHORIZATION
     tweet uses OAuth for authentication with Twitter.	This means that in
     order to run the application, it needs a so-called "consumer key" and
     "consumer secret".	 These are retrieved by registering an application
     with Twitter.

     Now for hopefully obvious reasons it's undesirable for an application to
     be distributed with its own key and secret; unfortunately that means that
     anybody wishing to run this tool would have to get their own consumer
     secret and key.

     Hence, if you wish to run this tool, please go to https://developer.twit-
     ter.com/ and register a new client.  Then store the key and secret in the
     configuration file (see below).

CONFIGURATION FILE
     tweet looks for user access information in ~/.tweetrc.  It will ignore
     all lines except for lines matching the patterns <user>_key=<key> and
     <user>_secret=<key>.  Those values need to be the access credentials cre-
     ated via OAuth (which tweet will generate for you and write to the file
     if not present for the given user).

     In addition to these, tweet also needs its own api credentials.  These
     are stored as (literally) "<api>_key" and "<api>_secret" (ie the username
     is "<api>", including the <>).

EXIT STATUS
     The tweet utility exits 0 on success, and >0 if an error occurs.

HISTORY
     tweet was originally written by Jan Schaumann <jschauma@netmeister.org>
     in January 2011.

BUGS
     Please report bugs and feature requests to the author.
```


CHANGES
Version 1.5 (2022-06-20)
 * move to python 3.x
 * add '-R' to un-retweet

Version 1.4 (2017-11-28)
 * use 280 chars
 * allow multi-line input

Version 1.3 (2016-07-25)
 * provide a more useful exit code

Version 1.2 (2016-07-06)
 * add '-m' to allow posting of media

Version 1.1 (2016-02-11)
 * add '-l' to fave^Wlike a given tweet

Version 1.0 (2015-03-08)
 * add '-c' to specify an alternate configuration file

Version 0.9 (2015-03-05)
 * add '-i' to print the ID of a new tweet

Version 0.8 (2012-04-18)
 * let's spell September correctly
 * spelling fixes in help (reported by @bcharder)
 * don't bother shortening links -- Twitter will just t.co wrap those further,
   possibly creating longer links

Version 0.7 (2011-09-06)
 * mention '-B' and '-b' in usage
 * add support for replying to a message by passing the ID via '-a'

Version 0.6 (2011-04-20)
 * add support for blocking and unblocking users

Version 0.5 (2011-03-15)
 * add support for deleting tweets

Version 0.4 (2011-03-13)
 * add support for following and unfollowing users
 * properly handle exceptions in the post-python-2.6 world

Version 0.3 (2011-02-16)
 * require user to be given via "-u"
 * allow for retweeting of messages given via "-r id" (multiple possible)

Version 0.2 (2011-02-15):
 * support '-t' flag to truncate long messages if desired

Version 0.1 (2011-01-10):
 * Initial version


src/tweet.py
#! /usr/bin/env python2.7
#
# This little tool is a very simple command-line tweeter.  That is, it
# allows you to send a twitter update to a given user account.  It doesn't
# do anything else.
#
# Copyright (c) 2011, Jan Schaumann. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#    1. Redistributions of source code must retain the above copyright notice,
#       this list of conditions and the following disclaimer.
#
#    2. Redistributions in binary form must reproduce the above copyright notice,
#       this list of conditions and the following disclaimer in the documentation
#       and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
# IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
# EVENT SHALL <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
# OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# Originally written by Jan Schaumann <jschauma@netmeister.org> in January 2011.

import getopt
import os
import re
import sys
import tweepy

###
### Classes
###

class Tweet(object):
    """A very simple twitter updater."""

    EXIT_ERROR = 1
    EXIT_SUCCESS = 0

    MAXCHARS = 280

    # https://dev.twitter.com/docs/tco-link-wrapper/faq
    # Technically, we should query and retrieve the max length for t.co
    # links, but I'm gonna go ahead and just use the current (as of
    # 2012-04-18) value.
    SHORT_LENGTH = 21

    def __init__(self):
        """Construct a Tweet object with default values."""

        self.__opts = {
                    "cfg_file"   : os.path.expanduser("~/.tweetrc"),
                    "blockees"   : [],
                    "aid"        : None,
                    "dids"       : [],
                    "friends"    : [],
                    "foes"       : [],
                    "likes"      : [],
                    "media"      : None,
                    "pids"       : False,
                    "rtids"      : [],
                    "truncate"   : False,
                    "unblockees" : [],
                    "user"       : ""
                 }
        self.auth = None
        self.api = None
        self.api_credentials = {}
        self.users = {}
        self.msg = ""
        self.we_tweet = True


    class Usage(Exception):
        """A simple exception that provides a usage statement and a return code."""

        def __init__(self, rval):
            self.err = rval
            self.msg = 'Usage: %s [-hit] [-[BF] user] [-[adlr] id] [-m media]\n' % os.path.basename(sys.argv[0])
            self.msg += '  [-[bf] user] -u user\n'
            self.msg += '\t-B user  unblock this user\n'
            self.msg += '\t-F user  unfollow this user\n'
            self.msg += '\t-a id    answer given message\n'
            self.msg += '\t-b user  block this user\n'
            self.msg += '\t-d id    delete given message\n'
            self.msg += '\t-f user  follow this user\n'
            self.msg += '\t-h       print this message and exit\n'
            self.msg += '\t-i       print the tweet ID of any new tweets\n'
            self.msg += '\t-l id    like given message\n'
            self.msg += '\t-m media attach media to the tweet\n'
            self.msg += '\t-r id    retweet given message\n'
            self.msg += '\t-t       truncate messages\n'
            self.msg += '\t-u user  tweet as this user\n'


    def block(self, blockees):
        """Block the given users."""

        for b in blockees:
            try:
                self.api.create_block(b)
            except tweepy.error.TweepError, e:
                sys.stderr.write("Error blocking %s: %s\n" % (f, e))
                return 1

        return 0


    def delete(self, ids):
        """Delete the given message."""

        for msg in ids:
            try:
                self.api.destroy_status(msg)
            except tweepy.error.TweepError, e:
                sys.stderr.write("Error deleting %s: %s\n" % (msg, e))
                return 1

        return 0

    def follow(self, friends):
        """Follow the given users."""

        for f in friends:
            try:
                self.api.create_friendship(f)
            except tweepy.error.TweepError, e:
                sys.stderr.write("Error following %s: %s\n" % (f, e))
                return 1

        return 0


    def favorite(self, ids):
        """Favorite^WLike the given tweets."""

        for msg in ids:
            try:
                self.api.create_favorite(msg)
            except tweepy.error.TweepError, e:
                sys.stderr.write("Error faving %s: %s\n" % (msg, e))
                return 1

        return 0


    def getAccessInfo(self, user):
        """Initialize OAuth Access Info (if not found in the configuration file)."""

        self.auth = tweepy.OAuthHandler(self.api_credentials['key'], self.api_credentials['secret'])

        if self.users.has_key(user):
            return

        auth_url = self.auth.get_authorization_url(True)
        print "Access credentials for %s not found in %s." % (user, self.getOpt("cfg_file"))
        print "Please log in on twitter.com as %s and then go to: " % user
        print "  " + auth_url
        verifier = raw_input("Enter PIN: ").strip()
        self.auth.get_access_token(verifier)

        self.users[user] = {
            "key" : self.auth.access_token,
            "secret" : self.auth.access_token_secret
        }

        cfile = self.getOpt("cfg_file")
        try:
            f = file(cfile, "a")
            f.write("%s_key = %s\n" % (user, self.auth.access_token))
            f.write("%s_secret = %s\n" % (user, self.auth.access_token_secret))
            f.close()
        except IOError, e:
            sys.stderr.write("Unable to write to config file '%s': %s\n" % \
                (cfile, e.strerror))
            raise


    def getLen(self, msg):
        """Return the length of the message, accounting for t.co link
        wrapping by Twitter."""

        pattern = re.compile('^(ftp|https?)://.+$')

        words = []
        length = 0

        for word in msg.split():
            if pattern.match(word):
                length += self.SHORT_LENGTH
            else:
                length += len(word)
            # whitespace
            length += 1

        return length


    def getOpt(self, opt):
        """Retrieve the given configuration option.

        Returns:
            The value for the given option if it exists, None otherwise.
        """

        try:
            r = self.__opts[opt]
        except ValueError:
            r = None

        return r


    def parseConfig(self, cfile):
        """Parse the configuration file and set appropriate variables.

        This function may throw an exception if it can't read or parse the
        configuration file (for any reason).

        Arguments:
            cfile -- the configuration file to parse

        Aborts:
            if we can't access the config file
        """

        try:
            f = file(cfile)
        except IOError, e:
            sys.stderr.write("Unable to open config file '%s': %s\n" % \
                (cfile, e.strerror))
            sys.exit(self.EXIT_ERROR)

        key_pattern = re.compile('^(?P<username>[^#]+)_key\s*=\s*(?P<key>.+)')
        secret_pattern = re.compile('^(?P<username>[^#]+)_secret\s*=\s*(?P<secret>.+)')
        for line in f.readlines():
            line = line.strip()
            key_match = key_pattern.match(line)
            if key_match:
                user = key_match.group('username')
                if user == "<api>":
                    self.api_credentials['key'] = key_match.group('key')
                else:
                    if self.users.has_key(user):
                        self.users[user]['key'] = key_match.group('key')
                    else:
                        self.users[user] = {
                            "key" : key_match.group('key')
                        }

            secret_match = secret_pattern.match(line)
            if secret_match:
                user = secret_match.group('username')
                if user == "<api>":
                    self.api_credentials['secret'] = secret_match.group('secret')
                else:
                    if self.users.has_key(user):
                        self.users[user]['secret'] = secret_match.group('secret')
                    else:
                        self.users[user] = {
                            "secret" : secret_match.group('secret')
                        }


    def parseOptions(self, inargs):
        """Parse given command-line options and set appropriate attributes.

        Arguments:
            inargs -- arguments to parse

        Raises:
            Usage -- if '-h' or invalid command-line args are given
        """

        try:
            opts, args = getopt.getopt(inargs, "B:F:a:b:c:d:f:hil:m:r:tu:")
        except getopt.GetoptError:
            raise self.Usage(self.EXIT_ERROR)

        for o, a in opts:
            if o in ("-B"):
                unblockees = self.getOpt("unblockees")
                unblockees.append(a)
                self.setOpt("unblockees", unblockees)
                self.we_tweet = False
            if o in ("-F"):
                foes = self.getOpt("foes")
                foes.append(a)
                self.setOpt("foes", foes)
                self.we_tweet = False
            if o in ("-a"):
                self.setOpt("aid", a)
            if o in ("-b"):
                blockees = self.getOpt("blockees")
                blockees.append(a)
                self.setOpt("blockees", blockees)
                self.we_tweet = False
            if o in ("-c"):
                self.setOpt("cfg_file", a)
            if o in ("-d"):
                dids = self.getOpt("dids")
                dids.append(a)
                self.setOpt("dids", dids)
                self.we_tweet = False
            if o in ("-h"):
                raise self.Usage(self.EXIT_SUCCESS)
            if o in ("-f"):
                friends = self.getOpt("friends")
                friends.append(a)
                self.setOpt("friends", friends)
                self.we_tweet = False
            if o in ("-i"):
                self.setOpt("pids", True)
            if o in ("-l"):
                likes = self.getOpt("likes")
                likes.append(a)
                self.setOpt("likes", likes)
                self.we_tweet = False
            if o in ("-m"):
                self.setOpt("media", a)
            if o in ("-r"):
                rtids = self.getOpt("rtids")
                rtids.append(a)
                self.setOpt("rtids", rtids)
                self.we_tweet = False
            if o in ("-t"):
                self.setOpt("truncate", True)
            if o in ("-u"):
                self.setOpt("user", a)

        if not self.getOpt("user") or args:
            raise self.Usage(self.EXIT_ERROR)


    def readInput(self):
        """Read the input from stdin.

        Arguments: none

        Returns:
            the given input, possibly modified

        Aborts:
            if given input is > MAXCHARS characters
        """

        lines = sys.stdin.readlines()
        if len(lines) < 2:
            msg = lines[0].strip()
        else:
            lines[len(lines)-1] = lines[len(lines)-1].strip()
            msg = "".join(lines)

        l=self.getLen(msg)
        if l > self.MAXCHARS:
            if self.getOpt("truncate"):
                truncLen = self.MAXCHARS - 4
                msg = ' '.join(msg[:truncLen].split(' ')[0:-1]) + '...'
            else:
                sys.stderr.write("Message too long (%d). Trim by %d.\n" \
                    % (l, (l - self.MAXCHARS)))
                sys.exit(self.EXIT_ERROR)

        return msg


    def retweet(self, ids):
        """Re-Tweet the given message."""

        for msg in ids:
            try:
                self.api.retweet(msg)
            except tweepy.error.TweepError, e:
                sys.stderr.write("Error retweeting %s: %s\n" % (msg, e))
                return 1

        return 0


    def setOpt(self, opt, val):
        """Set the given option to the provided value"""

        self.__opts[opt] = val


    def setupApi(self, user):
        """Create the object's api"""

        key = self.users[user]["key"]
        secret = self.users[user]["secret"]
        self.auth.set_access_token(key, secret)

        self.api = tweepy.API(self.auth)


    def tweet(self):
        """Read and then tweet the message"""

        status = None
        media = self.getOpt("media")

        msg = self.readInput()

        try:

            if media:
                if not os.access(media, os.F_OK):
                    sys.stderr.write("No such file: %s\n" % media)
                    return 1
                if not os.access(media, os.R_OK):
                    sys.stderr.write("Unable to read: %s\n" % media)
                    return 1

                status = self.api.update_with_media(media, status=msg, in_reply_to_status_id=self.getOpt("aid"))
            else:
                status = self.api.update_status(msg, self.getOpt("aid"))

            if self.getOpt("pids"):
                print status.id

        except tweepy.error.TweepError, e:
            sys.stderr.write("Unable to tweet: %s\n" % e)
            return 1

        return 0


    def unblock(self, unblockees):
        """Unblock the given users."""

        for b in unblockees:
            try:
                self.api.destroy_block(b)
            except tweepy.error.TweepError, e:
                sys.stderr.write("Error un-blocking %s: %s\n" % (f, e))
                return 1

        return 0


    def unfollow(self, foes):
        """Un-Follow the given users."""

        for f in foes:
            try:
                self.api.destroy_friendship(f)
            except tweepy.error.TweepError, e:
                sys.stderr.write("Error un-following %s: %s\n" % (f, e))
                return 1

        return 0


    def verifyConfig(self):
        """Verify that we have api credentials."""

        if (not (self.api_credentials.has_key("key") and self.api_credentials.has_key("secret"))):
            sys.stderr.write("No API credentials found.  Please do the 'register-this-app' dance.\n")
            sys.stderr.write("See 'man tweet' for more information.\n")
            sys.exit(self.EXIT_ERROR)


###
### "Main"
###

if __name__ == "__main__":
    try:
        tweet = Tweet()
        try:
            rval = 0
            tweet.parseOptions(sys.argv[1:])
            tweet.parseConfig(tweet.getOpt("cfg_file"))
            tweet.verifyConfig()

            user = tweet.getOpt("user")

            tweet.getAccessInfo(user)
            tweet.setupApi(user)

            if tweet.we_tweet:
                sys.exit(tweet.tweet())

            # Many of the things we can do involve getting a list and
            # passing it to a function.  Let's map these, so we can more
            # easily iterate over them:
            actions = [ ("blockees", tweet.block),
                            ("unblockees", tweet.unblock),
                            ("dids", tweet.delete),
                            ("friends", tweet.follow),
                            ("foes", tweet.unfollow),
                            ("likes", tweet.favorite),
                            ("rtids", tweet.retweet) ]

            for (thing, func) in actions:
                things = tweet.getOpt(thing)
                rval += func(things)

            sys.exit(rval)

        except tweet.Usage, u:
            if (u.err == tweet.EXIT_ERROR):
                out = sys.stderr
            else:
                out = sys.stdout
            out.write(u.msg)
            sys.exit(u.err)
            # NOTREACHED

    except KeyboardInterrupt:
        # catch ^C, so we don't get a "confusing" python trace
        sys.exit(Tweet.EXIT_ERROR)
